File: trees_and_graphs/number_of_connected_components_in_an_undirected_graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var countComponents = function(n, edges) {
    const dfsRecursive = (graph, seen, start) => {
        seen[start] = true;

        for (const neighbor of graph.get(start))
        {
            if (!seen[neighbor])
            {
                dfsRecursive(graph, seen, neighbor);
            }
        }
    }

    let components = 0;

    let visited = new Array(n).fill(false);
    let graph = new Map();

    for (let i = 0; i < n; i++)
    {
        graph.set(i, []);
    }

    for (const [x, y] of edges)
    {
        graph.get(x).push(y);
        graph.get(y).push(x);
    }

    for (let node = 0; node < n; node++)
    {
        if (!visited[node])
        {
            components++;
            dfsRecursive(graph, visited, node);
        }
    }

    return components;
};

File: trees_and_graphs/number_of_connected_components_in_an_undirected_graph.py
from collections import defaultdict

def countComponents(n: int, edges: list[list[int]]) -> int:
    def dfsRecursive(start):
        visited[start] = True

        for neighbor in graph[start]:
            if not visited[neighbor]:
                dfsRecursive(neighbor)

    if n == 0:
        return 0

    components = 0
    visited = [False] * n
    graph = defaultdict(list)

    for x, y in edges:
        graph[x].append(y)
        graph[y].append(x)
    
    for node in range(n):
        if not visited[node]:
            components += 1
            dfsRecursive(node)
    
    return components
